### ISSUE TITLE
Optimize availability of exports after bulk import

### ISSUE THREAD

**Issue by paynejd** [OpenConceptLab/ocl_issues#572]
The PEPFAR IMAP import script creates 1 org that consists of 1 source and ~100 collections. The IMAP export script retrieves all of these in parallel, does some ETL, and then returns. When requesting an IMAP export shortly after the bulk import from the IMAP import script is done processing, more than half of the collection exports do not yet exist, which causes the initial export to take >10 min while it waits for the exports to be cached on S3. Subsequent requests process much quicker.

In oclapi1, the first IMAP export request will occasionally have to wait for a few of the collection exports to be available, but not many, meaning that the first IMAP export request may take 20-30seconds longer than a later request.

Can we optimize the availability of cached exports so that the IMAP export script is never waiting more than a minute?

**Comment by paynejd** [comment]
@snyaggarwal FYI

**Comment by snyaggarwal** [comment]
@paynejd I have looked at the queries and added eager loading of relations:
 1. source/collection with owners
 2. concepts/mappings with parents and owners.
 This is a quick small improvement. Let me know if this improves the performance enough

**Comment by paynejd** [comment]
Hi Sunny, I just re-ran the scripts and this did not have any impact on the
performance

**Comment by paynejd** [comment]
could not test due to #595

**Comment by rkorytkowski** [comment]
ad3cf6e introduces a separate queue for indexing and a dedicated worker. Exports should complete much faster now as they will not wait in queue with indexing tasks.

I'll deploy it shortly to QA.

**Comment by rkorytkowski** [comment]
There's another issue with seed_concepts and seed_mappings which are highly innefficient. They need to fetch each concept/mapping in order to add it to the new version. I'll replace that with a bulk operation.

**Comment by snyaggarwal** [comment]
@rkorytkowski I have updated the logic to bulk index the concepts/mappings on new version creation https://github.com/OpenConceptLab/oclapi2/commit/dfe9215243deaa0ff7edd6dbf0aaed2dafd9f208

**Comment by rkorytkowski** [comment]
@snyaggarwal just made a commit to ignore processing flag when requesting an export if it is available. https://github.com/OpenConceptLab/oclapi2/commit/ee5838aa3654c67e6e7275b8f3cffc22cd257bde

@snyaggarwal could you please make sure to start a commit message with OpenConceptLab/ocl_issues#572 so it's automatically linked to a GH issue? (in the future)

**Comment by rkorytkowski** [comment]
We need to push that to staging so @paynejd can test and we can close this one.

**Comment by paynejd** [comment]
The changes committed improved performance significantly, but still a ways to go. The initial export of an IMAP (consisting of 100+ collections) still took 4 minutes (down from >10min), while a subsequent export took only 26 seconds. I've attached the log output of the export here.

[staging.aws-BI-FY19-baseline-export.log](https://github.com/OpenConceptLab/ocl_issues/files/6031949/staging.aws-BI-FY19-baseline-export.log)

@snyaggarwal @rkorytkowski Do you have recommendations to get the initial export time down to under a minute? If not, @snyaggarwal is it time to look at either exporting a full org in one request or refactoring the script to use direct queries?

**Comment by rkorytkowski** [comment]
@paynejd I need to check SQL queries... I can look into that on Friday.

**Comment by paynejd** [comment]
This is now working perfectly! A script that was taking ~15min to run is now taking ~55sec. Thank you @rkorytkowski and @snyaggarwal!
